### PR TITLE
[Github CRUD] CRUD Wrapper for Github Repositories

### PR DESCRIPTION
**Pull request by TAMREF**
Design doc: [notion](https://www.notion.so/youngyojun/GitHub-Repository-CRUD-5a5181e2f21c4919b2f7eb4ec4b8b506?pvs=4)

- [x] Naive CRUD wrapper
- [x] CRUD based on existence of `polytope.yaml`
- [x] tests

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -161,3 +161,8 @@ cython_debug/
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
 #.idea/
+
+
+# File for internal testing
+src/polytope/main_test.py
+src/polytope/main_test.sh
```
Comment (youngyojun) [inline .gitignore:168]: (question) Are they necessary?

File: src/polytope/github/repository/InternalCode.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class GithubRepositoryInternalCode(Enum):
+    """! Internal status code of Github CRUD requests."""
+
+    # Success.
+    Success = auto()
+    # create cannot modify repository name.
+    ForbiddenNameChangeOnCreation = auto()
+    # config validation failed.
```
Comment (youngyojun) [inline src/polytope/github/repository/InternalCode.py]: (suggestion) No need to do so, but please match the capitalization of the first word in each comment. (`Success` -> `success`)

File: src/polytope/github/repository/InternalCode.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    # creation (without template) failed on Github side.
+    FailedToCreateWithoutTemplate = auto()
+    # update failed on Github side.
+    FailedToUpdate = auto()
+    # deletion failed on Github side.
+    FailedToDelete = auto()
+    # cannot update repository without polytope config file.
+    UpdateWithoutPolytopeFile = auto()
+    # cannot delete repository without polytope config file.
+    DeleteWithoutPolytopeFile = auto()
```
Comment (youngyojun) [inline src/polytope/github/repository/InternalCode.py]: (suggestion) Please always capitalize `Polytope`.

File: src/polytope/github/repository/Repository.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    Controller of Github Repository.
+
+    @param token: Personal Access Token.
+    @param owner: Github Username of Repository Owner.
+    @param name: Name of the repository.
+    """
+
+    def __init__(
+        self, owner: str, name: str, token: "Token", session: Session = RequestsSession
+    ) -> None:
```
Comment (youngyojun) [inline src/polytope/github/repository/Repository.py]: (comment) Please use Python Doxygen comment style.

File: src/polytope/github/repository/Repository.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+        # Much generous validation than Github's.
+        # Github has 39 chars for username at max.
+        assert 0 < len(owner)
+        assert is_valid_github_user_name(owner)
+        # TODO name validation?
```
Comment (youngyojun) [inline src/polytope/github/repository/Repository.py]: You are already implementing name validation?

File: src/polytope/github/repository/Repository.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+    """
+    Controller of Github Repository.
+
+    @param token: Personal Access Token.
+    @param owner: Github Username of Repository Owner.
+    @param name: Name of the repository.
+    """
+
+    def __init__(
+        self, owner: str, name: str, token: "Token", session: Session = RequestsSession
```
Comment (youngyojun) [inline src/polytope/github/repository/Repository.py]: (comment) The type of `session` should be `Type[Session]` not `Session`, I think.

File: src/polytope/github/repository/InternalCode.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    """! Internal status code of Github CRUD requests."""
+
+    # Success.
+    Success = auto()
+    # create cannot modify repository name.
+    ForbiddenNameChangeOnCreation = auto()
+    # config validation failed.
+    ConfigValidationFailed = auto()
+    # creation failed on Github side.
+    FailedToCreate = auto()
```
Comment (youngyojun) [inline src/polytope/github/repository/InternalCode.py]: (question) Should we distinguish the case of creation failed due to duplication of repositories' names?

**Review by youngyojun — COMMENTED** [review]
Awesome! Thank you for your hard work! :tada: Here are some of my comments.

File: .gitignore (hunk 1 above)
Comment (TAMREF) [reply .gitignore:168]: I mainly used them for HTTP test. We don't have to keep it, but global test dir/file could be useful.

File: src/polytope/github/repository/InternalCode.py (hunk 7 above)
Comment (TAMREF) [reply src/polytope/github/repository/InternalCode.py]: `FailedToXXX` is the most broad term of failure. It'll be always good if we can elaborate the reason, at some rational cost.
For your suggestion, we can do it by
1. Parse Github error response.
2. Or make a READ call in advance.

File: src/polytope/github/repository/Repository.py (hunk 5 above)
Comment (TAMREF) [reply src/polytope/github/repository/Repository.py]: I'll remove TODO. Thanks for the heads up.

File: src/polytope/github/repository/InternalCode.py (hunk 7 above)
Comment (TAMREF) [reply src/polytope/github/repository/InternalCode.py]: I'll lease this for future with a todo mark. It'll be efficient to focus on frequent and valuable reasons of failure.

**Review by youngyojun — APPROVED** [review]
Thank you for considering and reflecting the review!
Looks good to me 👍

**Review by utilForever — COMMENTED** [review]
I forgot this PR. LGTM. Thanks! 🗡️

**Pull request merged**